docs/src/constraints.md

```diff
@@ -1,5 +1,531 @@
+true
+
+julia> JuMP.dual(con)
+-2.0
+
+julia> @objective(model, Max, 2x)
+2 x
+
+julia> JuMP.optimize!(model)
+
+julia> JuMP.dual(con)
+-2.0
+```
+
+To help users who may be less familiar with conic duality, JuMP provides the
+[`JuMP.shadow_price`](@ref) function which returns a value that can be
+interpreted as the improvement in the objective an infinitesimal relaxation (on
+the scale of one unit) in the right-hand side of the constraint.
+[`JuMP.shadow_price`](@ref) can only be used on linear constraints with a `<=`,
+`>=`, or `==` comparison operator.
+
+In the example above, `JuMP.dual(con)` returned `-2.0` regardless of the
+optimization sense. However, in the second case when the optimization sense is
+`Max`, [`JuMP.shadow_price`](@ref) returns:
+```jldoctest con_duality
+julia> JuMP.shadow_price(con)
+2.0
+```
+
+To query the dual variables associated a variable bound, first obtain a
+constraint reference using one of [`JuMP.UpperBoundRef`](@ref),
+[`JuMP.LowerBoundRef`](@ref), or [`JuMP.FixRef`](@ref), and then call
+[`JuMP.dual`](@ref) on the returned constraint reference.
+
+```@meta
+DocTestSetup = quote
+    using JuMP
+    const MOI = JuMP.MathOptInterface
+end
+```
+
+## Constraint containers
+
+So far, we've added constraints one-by-one. However, just like
+[Variable containers](@ref), JuMP provides a mechanism for building groups of
+constraints compactly. References to these groups of constraints are returned in
+*containers*. Three types of constraint containers are supported: `Array`s,
+`JuMPArray`s, and `Dict`ionaries. We explain each of these in the following.
+
+### [Arrays](@id constraint_arrays)
+
+One way of adding a group of constraints compactly is the following:
+```jldoctest constraint_arrays; setup=:(model=Model(); @variable(model, x))
+julia> @constraint(model, con[i = 1:3], i * x <= i + 1)
+3-element Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,1}:
+ con[1] : x <= 2.0
+ con[2] : 2 x <= 3.0
+ con[3] : 3 x <= 4.0
+```
+JuMP returns references to the three constraints in an `Array` that is bound to
+the Julia variable `con`. This array can be accessed and sliced just like a
+normal Julia array:
+```jldoctest constraint_arrays
+julia> con[1]
+con[1] : x <= 2.0
+
+julia> con[2:3]
+2-element Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,1}:
+ con[2] : 2 x <= 3.0
+ con[3] : 3 x <= 4.0
+```
+
+Anonymous containers can also be constructed by dropping the name (e.g. `con`)
+before the square brackets:
+```jldoctest constraint_arrays
+julia> @constraint(model, [i = 1:2], i * x <= i + 1)
+2-element Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,1}:
+ x <= 2.0
+ 2 x <= 3.0
+```
+
+Just like [`@variable`](@ref), JuMP will form an `Array` of constraints when it
+can determine at compile time that the indices are one-based integer ranges.
+Therefore `con[1:b]` will create an `Array`, but `con[a:b]` will not. If JuMP
```

Comment:
It is unclear why for `1:b`, we can see at compile time that it is one-based because it has the same type than `2:b`. Maybe "parsing time" instead of "compile time" ?

Reply:
In https://docs.julialang.org/en/v1/manual/metaprogramming/#Hold-up:-why-macros?-1, they say

> Macros are necessary because they execute when code is parsed

so it seems that "parse" is accurate